### DISCUSSION TITLE
Self hosting guidelines

### DISCUSSION THREAD

**Discussion by apperside** [nhost/nhost#970] [General]
Hi guys,
first of all congratulations for this awesome project!
I am playing around with self hosting and it is really really easy and works like a charm! However, the docs about self hosting are quite minimal, there is just the example project, but I'd like to have some suggestions about going in production with the self hosted version:

Is there anything I need to care about after being sure that the secrets in the docker configuration file are not exposed? 

Are there other limitations except that you can't use the nhost dashboard? 

When there is a new nhost version, how do I update it?

Many thanks in advance 🙏

**Comment by T-travis** [comment]
An in-depth user guide for self hosting with prod recommended settings would be awesome.

**Reply by dbarrosop** [reply]
The provided [docker-compose](https://github.com/nhost/nhost/tree/main/examples/docker-compose) is only for demonstration purposes. The README.md is actually quite clear on that.

And the CLI is not meant to be for hosting purposes and we don't support that use-case. It is specifically designed to provide a development environment suitable to integrate with our cloud.

**Reply by mohamed-ziny** [reply]
@dbarrosop And there are no plans to fix this mistake?

**Comment by elitan** [comment]
Yes, we need to improve the docs for self-hosting.

The best resource, for now, is our example here: https://github.com/nhost/nhost/tree/main/examples/docker-compose

**Comment by foodelix** [comment]
Hello @elitan ,
thanks for the update. Is there any plan for it or is it something that is not planned for now?

Thank you

**Comment by R3flector** [comment]
after docker compose up. graphql-engine loop restarting

**Comment by superstructor** [comment]
Would be great if we could get more clarity on self-hosting, esp how to setup GitHub deployments like can be done on cloud.

**Comment by nokia6290** [comment]
Any progress here?

**Comment by GFoley83** [comment]
@dbarrosop could you please weigh in on this one? 

Seems like there's no progress here beyond looking at the docker-compose. At the very least, some guidance on production configuration and recommended cloud services offerings for the various features (outside of running docker up on an EC2 or deploying everything to EKS), would be welcome.

**Reply by dbarrosop** [reply]
> and recommended cloud services offerings for the various features

Of course, we recommend using https://nhost.io :) Using our cloud means supporting the team building the software you are using.

Other than that, no, we don't have any other recommendation. Self-hosted is community-supported unless you sign up for a support agreement.

**Reply by GFoley83** [reply]
Thanks @dbarrosop. 

Of course, you want everyone to use you'd paid offering, which is perfectly reasonable. But for people like myself who are initially only interested in self-hosting to get started, having little to no docs or official support is not a great initial experience. 
I'd argue that your users who are self-hosting are probably a lot more likely to contribute & improve your code-base than your paying customers. It's hard to do that though when even your docker-compose example has issues. 

In any case, if I could get off the starter block, I'd be more than happy to contribute; much like I did for [Activepieces ](https://github.com/activepieces/activepieces) when I wrote their deploy to AWS docs.

**Comment by zuzu-lulu** [comment]
man wtf i cant self host because they want you to pay for it

**Reply by dbarrosop** [reply]
You can self-host, you just need self-hosting skills. We even have a 100% working example in our repo that takes two commands to get running.

**Reply by zuzu-lulu** [reply]
then why is everyone here and outside saying its impossible/very difficult to self-host ?

**Reply by dbarrosop** [reply]
I don't know, it is not my place to speculate. But not having the skills to self-host/knowing how to do it isn't the same as not being able to do it. Proof that it can be self-hosted is our cli, which runs an nhost project on your machine and our example:

https://github.com/nhost/nhost/tree/main/examples/docker-compose

If people wants to self-host I'd expect them to be able to understand at least a docker-compose file and being able to map it to their environment. If people don't have the skills and they still want/need to self-host we have various support models/options for them:

https://docs.nhost.io/platform/self-hosting/overview

**Reply by reynaldi-lorenzo** [reply]
i have successfully working on in too. But now, i want self-hosted support multi-project also. will try to read more on your graphql schema so i can mimicing central config server also.

**Comment by acamarata** [comment]
I just made nself.org for this exact reason

**Comment by myagizmaktav** [comment]
# Setting Up Local Domains for Nhost

To access the Nhost services in your browser, you need to add local domain mappings to your system's hosts file.

## Quick Setup

### macOS / Linux

1. **Open Terminal**

2. **Run this single command** (it will ask for your password):

```bash
sudo bash -c 'cat >> /etc/hosts << EOF

# Nhost Docker Compose local domains
127.0.0.1 local.auth.local.nhost.run
127.0.0.1 local.console.local.nhost.run
127.0.0.1 local.dashboard.local.nhost.run
127.0.0.1 local.db.local.nhost.run
127.0.0.1 local.functions.local.nhost.run
127.0.0.1 local.graphql.local.nhost.run
127.0.0.1 local.mailhog.local.nhost.run
127.0.0.1 local.storage.local.nhost.run
EOF
'
```

3. **Done!** You can now access the services in your browser.

---

### Windows

1. **Open Notepad as Administrator**
   - Press `Windows Key`
   - Type "Notepad"
   - Right-click on Notepad
   - Select "Run as administrator"

2. **Open the hosts file**
   - Click `File` → `Open`
   - Navigate to: `C:\Windows\System32\drivers\etc`
   - Change file type filter to "All Files (*.*)"
   - Open the file named `hosts`

3. **Add these lines at the end of the file:**

```
# Nhost Docker Compose local domains
127.0.0.1 local.auth.local.nhost.run
127.0.0.1 local.console.local.nhost.run
127.0.0.1 local.dashboard.local.nhost.run
127.0.0.1 local.db.local.nhost.run
127.0.0.1 local.functions.local.nhost.run
127.0.0.1 local.graphql.local.nhost.run
127.0.0.1 local.mailhog.local.nhost.run
127.0.0.1 local.storage.local.nhost.run
```

4. **Save and close Notepad**

5. **Flush DNS cache** (open Command Prompt as Administrator):
```cmd
ipconfig /flushdns
```

---

## Manual Setup (Alternative)

If you prefer to edit the file manually:

### macOS / Linux

1. Open Terminal

2. Edit the hosts file:
```bash
sudo nano /etc/hosts
```

3. Add the lines from above at the end of the file

4. Save and exit:
   - Press `Ctrl + O` to save
   - Press `Enter` to confirm
   - Press `Ctrl + X` to exit

---

## Verify It's Working

After adding the entries, test the services:

```bash
# Test Auth
curl http://local.auth.local.nhost.run/v1/version

# Test Functions
curl http://local.functions.local.nhost.run/v1/hello

# Test Storage
curl http://local.storage.local.nhost.run/v1/version

# Test GraphQL
curl http://local.graphql.local.nhost.run/v1/version
```

Or simply open these URLs in your browser:

- **Dashboard**: http://local.dashboard.local.nhost.run
- **GraphQL Console**: http://local.graphql.local.nhost.run/console
- **Mailhog (Email Testing)**: http://local.mailhog.local.nhost.run

---

## Troubleshooting

### Still can't access the services?

1. **Check Docker is running:**
```bash
docker compose ps
```
All services should show as "running".

2. **Verify hosts file was updated:**
```bash
# macOS / Linux
cat /etc/hosts | grep nhost

# Windows (in Command Prompt)
type C:\Windows\System32\drivers\etc\hosts | findstr nhost
```

3. **Flush DNS cache:**

**macOS:**
```bash
sudo dscacheutil -flushcache
sudo killall -HUP mDNSResponder
```

**Linux:**
```bash
sudo systemd-resolve --flush-caches
# or
sudo /etc/init.d/nscd restart
```

**Windows:**
```cmd
ipconfig /flushdns
```

4. **Try a different browser or incognito/private mode**

5. **Restart Docker Compose:**
```bash
docker compose down
docker compose up
```

---

## Remove Entries Later

If you want to remove these entries later:

### macOS / Linux
```bash
sudo nano /etc/hosts
# Delete the lines that start with "# Nhost Docker Compose" and the 8 lines below it
```

### Windows
1. Open Notepad as Administrator
2. Open `C:\Windows\System32\drivers\etc\hosts`
3. Delete the Nhost lines
4. Save the file
